sources/api/shibaken/src/main.rs

```diff
@@ -0,0 +1,278 @@
+                &public_key_text.trim_end()
+            );
+            continue;
+        }
+    }
+    ensure!(!public_keys.is_empty(), error::NoValidImdsKeys);
+
+    Ok(public_keys)
+}
+
+/// Returns a list of endpoint urls strings for the public keys in IMDS. Since IMDS returns the list
+/// of available public keys as '0=my-public-key', we need to strip the index from the list and
+/// embed it into the endpoint url.
+fn build_public_key_endpoints_from_imds(public_key_list: &str) -> Result<Vec<String>> {
+    let mut public_key_endpoints = Vec::new();
+    for available_key in public_key_list.lines() {
+        let f: Vec<&str> = available_key.split('=').collect();
+        // If f[0] isn't a number, then it isn't a valid index.
+        if f[0].parse::<i32>().is_ok() {
+            let public_key_endpoint = format!("{}{}/openssh-key", IMDS_PUBLIC_KEY_BASE_URL, f[0]);
+            public_key_endpoints.push(public_key_endpoint);
+        } else {
+            warn!(
+                "'{}' does not appear to be a valid index. Skipping...",
+                &f[0]
+            );
+            continue;
+        }
+    }
+    ensure!(
+        !public_key_endpoints.is_empty(),
+        error::NoValidImdsKeyEndpoints
+    );
+    Ok(public_key_endpoints)
+}
+
+/// Store the args we receive on the command line.
+struct Args {
+    log_level: LevelFilter,
+}
+
+/// Print a usage message in the event a bad arg is passed.
+fn usage() -> ! {
+    let program_name = env::args().next().unwrap_or_else(|| "program".to_string());
+    eprintln!(
+        r"Usage: {}
+            [ --log-level trace|debug|info|warn|error ]",
+        program_name
+    );
+    process::exit(2);
+}
+
+/// Prints a more specific message before exiting through usage().
+fn usage_msg<S: AsRef<str>>(msg: S) -> ! {
+    eprintln!("{}\n", msg.as_ref());
+    usage();
+}
+
+/// Parse the args to the program and return an Args struct.
+fn parse_args(args: env::Args) -> Args {
+    let mut log_level = None;
+
+    let mut iter = args.skip(1);
+    while let Some(arg) = iter.next() {
+        match arg.as_ref() {
+            "--log-level" => {
+                let log_level_str = iter
+                    .next()
+                    .unwrap_or_else(|| usage_msg("Did not give argument to --log-level"));
+                log_level = Some(LevelFilter::from_str(&log_level_str).unwrap_or_else(|_| {
+                    usage_msg(format!("Invalid log level '{}'", log_level_str))
+                }));
+            }
+
+            _ => usage(),
+        }
+    }
+
+    Args {
+        log_level: log_level.unwrap_or_else(|| LevelFilter::Error),
+    }
+}
+
+fn run() -> Result<()> {
+    let args = parse_args(env::args());
+
+    // TerminalMode::Mixed will send errors to stderr and anything less to stdout.
+    TermLogger::init(args.log_level, LogConfig::default(), TerminalMode::Mixed)
+        .context(error::Logger)?;
+
+    info!("shibaken started");
```

Comment:
Are these `info!` uses correct? It seems odd to mix `info!` and `println!` if sundog is expecting structured data on stdout.

Reply:
It's a tad tricky. shibaken is called by sundog which is expecting a JSON answer on **stdout**. It seems like it would be nice if this dog had a logging mechanism, but in `TerminalMode::Mixed` anything less than error would go to **stdout**. One alternative would be to set `TerminalMode::Stderr` which would send all logs to **stderr**, but it feels a little wrong. In this case, the default log level is set to Error, but then there isn't much point to having the info and warn logs. Would it be better to remove the info logs all together and send warn/error to **stderr** or keep the logging as is and send it all to **stderr**?

Reply:
Yeah, this is a bit of a new situation for us.  None of the other setting generators log anything precisely for this reason.

However, since this generator is mission-critical it seems like it should probably log what it's doing to make for easier troubleshooting in the event something bad happens.  It _is_ weird to log everything to stderr, but we don't have another option as far as I know.  We could remove all logging, but that seems like a poor option as well.

Reply:
>  None of the other setting generators log anything precisely for this reason

I think we just weren't thinking about it.  We should probably log to stderr.  All generators are mission critical!  :)